lib/Sema/SemaBounds.cpp

```diff
@@ -578,6 +578,79 @@ namespace {
 
       return true;
     }
+
+
+    bool VisitImplicitCastExpr(ImplicitCastExpr *E) {
+      if (E->getCastKind() == CK_BitCast)
+        FindDisallowedFunctionPtrCasts(E);
+
+      return true;
+    }
+
+  private:
+    // Here we're looking for places where the programmer has cast from an
+    // unchecked function pointer type to a checked function pointer type.
+    // We need to descend through all the casts below the top-level bitcast to find
+    // the eventual thing we're referring to. We then check to make sure that it is
+    // either a top-level function declaration (of either checked or unchecked type), or
+    // a local variable that has a checked function type.
+    //
+    // The only other exception we have is that we can always assign the null pointer to
+    // a checked function pointer type.
+    void FindDisallowedFunctionPtrCasts(ImplicitCastExpr *E) {
```

Comment:
We want this to apply to explicit user casts as well to checked pointer types, not just implicit casts.   So if a user writes:

ptr<int (void)> f = (ptr<int>(void)) (random_value)

this should be rejected.